lib/memo_wise.rb

```diff
@@ -114,6 +114,163 @@ def self.has_kwarg?(method) # rubocop:disable Naming/PredicateName
+    end
+
+    unless method_name.is_a?(Symbol)
+      raise ArgumentError, "#{method_name.inspect} must be a Symbol"
+    end
+
+    # In Ruby, "class methods" are implemented as normal instance methods
+    # on the "singleton class" of a given Class object, found via
+    # {Class#singleton_class}.
+    # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+    singleton = klass.singleton_class
+
+    is_singleton_method =
+      singleton.private_instance_methods.include?(method_name) ||
+      singleton.instance_methods.include?(method_name)
+
+    if is_singleton_method
+      visibility = method_visibility(singleton, method_name)
+      [:class, visibility, singleton, singleton.instance_method(method_name)]
+    else
+      visibility = method_visibility(klass, method_name)
+      [:instance, visibility, klass, klass.instance_method(method_name)]
+    end
+  end
+
+  # @private
+  #
+  # Returns visibility of an instance method defined on a class.
+  #
+  # @param klass [Class]
+  #   Class in which to find the visibility of an existing *instance* method.
+  #
+  # @param method_name [Symbol]
+  #   Name of existing *instance* method find the visibility of.
+  #
+  # @return [:private, :protected, :public]
+  #   Visibility of existing instance method of the class.
+  #
+  # @raise ArgumentError
+  #   Raises `ArgumentError` unless `method_name` is a `Symbol` corresponding
+  #   to an existing **instance** method defined on `klass`.
+  #
+  def self.method_visibility(klass, method_name)
+    if klass.private_method_defined?(method_name)
+      :private
+    elsif klass.protected_method_defined?(method_name)
+      :protected
+    elsif klass.public_method_defined?(method_name)
+      :public
+    else
+      raise ArgumentError, "#{method_name.inspect} must be a method on #{klass}"
+    end
+  end
+
+  # @private
+  #
+  # Find the original class for which the given class is the corresponding
+  # "singleton class".
+  #
+  # See https://stackoverflow.com/questions/54531270/retrieve-a-ruby-object-from-its-singleton-class
+  #
+  # @param klass [Class]
+  #   Singleton class to find the original class of
+  #
+  # @return Class
+  #   Original class for which `klass` is the singleton class.
+  #
+  # @raise ArgumentError
+  #   Raises if `klass` is not a singleton class.
+  #
+  def self.original_class_from_singleton(klass)
+    unless klass.singleton_class?
+      raise ArgumentError, "Must be a singleton class: #{klass.inspect}"
+    end
+
+    # Option 1: Derive from `#to_s` of the singleton class
+    #   * Works unless original class NOT assigned to constant
+    #   * Depends on an undocumented naming convention in Ruby, where singleton
+    #     class of `Foo` will return "#<Class:Foo>" from `#to_s` method.
+    name = klass.to_s.match(/#<Class:([^>]+)>/)&.[](1)
+    raise NameError if name.nil?
+
+    Object.const_get(name) # raises NameError if not found
```

Comment:
Is there a risk of this finding something but it not being the class we expect? We could mitigate this with:

```
Object.const_get(name).tap do |original|
  raise NameError unless original.singleton_class == klass
end
```

or just remove the `Option 1` approach entirely? Especially if this runs at load time I wonder if it's really worth it to optimize this so much, but I could be convinced either way.